[dtensor][TP] check funcol calls and improve doc for loss parallel

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121366

Since CommDebugMode is fixed, we can check that loss parallel is working as expected.

Under loss parallel, the forward computation should invoke 3 all-reduces, and the backward computation should invoke no functional collectives.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang

tianyu-l committed:
[dtensor][TP] check funcol calls and improve doc for loss parallel

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/loss.py:
```diff
@@ -26,10 +26,13 @@
+
+    Within this context manager, one can use :func:`~torch.nn.functional.cross_entropy` or
+    :class:`~torch.nn.CrossEntropyLoss` as usual, with the following assumptions on the input parameters.
+    The corresponding backward call, if any, also needs to happen under this context manager.
```
```suggestion
    The corresponding ``backward()`` call, if any, also needs to happen under this context manager.
```

wanchaol reviewed (approved):
lgtm!

tianyu-l committed:
Update torch/distributed/tensor/parallel/loss.py

Co-authored-by: Wanchao <wanchaol@users.noreply.github.com>

tianyu-l committed:
Update on "[dtensor][TP] check funcol calls and improve doc for loss parallel"


Since CommDebugMode is fixed, we can check that loss parallel is working as expected.

Under loss parallel, the forward computation should invoke 3 all-reduces, and the backward computation should invoke no functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge

tianyu-l committed:
Update on "[dtensor][TP] check funcol calls and improve doc for loss parallel"


Since CommDebugMode is fixed, we can check that loss parallel is working as expected.

Under loss parallel, the forward computation should invoke 3 all-reduces, and the backward computation should invoke no functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
